Update definition of Unicode Whitespace

amuino opened the pull request:
We have been using BiDi character types to determine which unicode characters to consider as whitespace (looking at `B`, `S` and `WS`). By some reason the character set of BiDi whitespace is different than the set for the unicode `White_Space` property. This difference causes surprise in users since the docs for `String.strip/1` and `String.split/1` refer to "unicode whitespace" (the main concern expressed so far is the exclusion of the no-break space `0x0020` from the BiDi set, see #4300).

This PR swaps the use of BiDi character types by the use of the `White_Space` property by adding a subset of a new Unicode table (`PropList.txt`) and a parser for it. This makes the implementation better match the documentation.

Additionally, the no-break characters are excluded from `split` (since splitting on a no-break character seems contradictory) and its documentation is updated to document this.

whatyouhide commented on lib/elixir/test/elixir/string_test.exs:
```diff
@@ -33,8 +33,12 @@ defmodule StringTest do
     assert String.split("foo\t\n\v\f\r\sbar\n") == ["foo", "bar"]
-    assert String.split("foo" <> <<31>> <> "bar") == ["foo", "bar"]
     assert String.split("foo" <> <<194, 133>> <> "bar") == ["foo", "bar"]
+    # information separators not considered whitespace
```
> information separators **are** not considered whitespace

Wdyt? :)

whatyouhide commented on lib/elixir/unicode/unicode.ex:
```diff
@@ -12,11 +12,25 @@ defmodule String.Unicode do
   def version, do: {7, 0, 0}
 
+  prop_path = Path.join(__DIR__, "PropList.txt")
+  prop_range_line = ~r/\A([0-9A-F]{4})(?:\.\.([0-9A-F]{4}))?/
```
Maybe we can call this `prop_range_line_regex`?

whatyouhide commented on lib/elixir/unicode/unicode.ex:
```diff
@@ -12,11 +12,25 @@ defmodule String.Unicode do
+        [to_binary.(single) | acc]
+      [_, first, last] ->
+        [first_i, last_i] = [String.to_integer(first, 16), String.to_integer(last, 16)]
+        acc ++ Enum.map(first_i..last_i, &(to_binary.(Integer.to_string &1, 16)))
```
You can drop the `(` after `&` here (and personally I'd add parens to the `Integer.to_string/2` call :bowtie:):

```elixir
acc ++ Enum.map(first_i..last_i, &to_binary.(Integer.to_string(&1, 16)))
```

amuino commented:
@whatyouhide added a new commit with the suggested improvements. Thanks for the quick review!

lexmag commented on lib/elixir/lib/string.ex:
```diff
@@ -241,7 +241,8 @@ defmodule String do
   occurrence with leading and trailing whitespace ignored. Groups
-  of whitespace are treated as a single occurrence.
+  of whitespace are treated as a single occurrence. Divisions do
+  not occur on whitespace marked as no-break by unicode.
```
Maybe we can say "on whitespace marked as no-break by unicode" –> "non-breakable whitespace"?

josevalim replied:
Don't forget to add `~S` to the sigil above to ensure `\u....` in the example below won't be escaped.

lexmag commented on lib/elixir/test/elixir/string_test.exs:
```diff
@@ -33,8 +33,12 @@ defmodule StringTest do
     assert String.split("foo" <> <<194, 133>> <> "bar") == ["foo", "bar"]
+    # information separators are not considered whitespace
+    assert String.split("foo\u001Fbar") == ["foo\u001Fbar"]
+    # no-break space is excluded
```
"no-break" –> "non-breaking"?

trejkaz replied:
Oof. It's this sort of thing which makes us app developers have to specifically add U+00A0 to the list of things to break on.

josevalim replied:
@trejkaz as you probably know, this is actually more complex because there are cases where you don't want to split on U+00A0. For example, as an example provided by Unicode itself, it can be used to separate numbers: `$3 567,00`.

The truth is that this `split` implementation is arbitrary and it will only work on the simplest of cases. For anything that actually requires understanding the text semantics, such as line/word boundary breaking, then you need to follow the rules in both Unicode and CLDR.

kipcole9 replied:
As @josevalim says, U+00A0 is specifically intended as a non-breaking space. As are U+202F, U+2007 and U+2060.

trejkaz replied:
But nonetheless a space, and thus we end up wanting to break on it anyway, if doing real text processing. But yeah, it is more complex than that too. Actually in the case of numbers you can have a breaking space there too, and people expect it to be processed as a number. If you have two words separated by a non-breaking space, they also expect it to be processed the same as a space. *shrug*

josevalim replied:
That's exactly the point. If you are doing real text processing, then you surely don't want to use `String.split`, regardless if it includes breaking or non-breaking spaces. Adding no-break spaces won't make it any more suitable for processing text. Not even passing your own list of characters would make it any better.

josevalim commented on lib/elixir/unicode/unicode.ex:
```diff
@@ -30,8 +44,8 @@ defmodule String.Unicode do
-      bidi in ["B", "S", "WS"] ->
-        {cacc, [to_binary.(codepoint) | wacc]}
+      String.starts_with?(decomposition, "<noBreak>") ->
+        {cacc, Enum.reject(wacc, &(&1 == to_binary.(codepoint)))}
```
`List.delete(wacc, to_binary.(codepoint))`

josevalim commented on lib/elixir/unicode/PropList.txt:
```diff
@@ -0,0 +1,27 @@
+
+# ================================================
+
+0009..000D    ; White_Space # Cc   [5] <control-0009>..<control-000D>
```
Can we remove all comments from this file and call it `WhiteSpace.txt`? My goal is to remove the regex from unicode.ex file. :) Then in the code itself we can comment it was extracted from PropList.txt. :)

amuino replied:
I assumed that the files where there to be easy to upgrade, so I barely touched it. But if we do not need the original files, what do you think about embedding the char list in the code itself (and keeping a comment to the source)? It is just 26 chars, IIRC.

josevalim replied:
It is fine so let's keep it as separate file still because we'll copy the section as is when there is a new unicode release.

amuino commented on lib/elixir/test/elixir/string_test.exs:
```diff
@@ -33,8 +33,12 @@ defmodule StringTest do
     assert String.split("foo" <> <<194, 133>> <> "bar") == ["foo", "bar"]
+    # information separators are not considered whitespace
+    assert String.split("foo\u001Fbar") == ["foo\u001Fbar"]
+    # no-break space is excluded
```
Weird as it sounds, the unicode name for `u+00A0` is _no-break whitespace_ ¯\_(ツ)_/¯ ([link](http://unicode.org/cldr/utility/list-unicodeset.jsp?a=%5Cp%7Bwhitespace%7D))

lexmag replied:
Oh, right, if it is the name for it we should keep it as is. :smile:

josevalim commented:
Beautiful @amuino! I have added some comments!

lexmag commented on lib/elixir/test/elixir/string_test.exs:
```diff
@@ -187,6 +190,10 @@ defmodule StringTest do
     assert String.rstrip("   cat 猫猫", ?猫) == "   cat "
+    # information separators are not whitespace 
+    assert String.rstrip("a  abc  a \u001F") == "a  abc  a \u001F"
+    # no-break whitespace
```
Then it is `no-break whitespace` –> `no-break space`.

lexmag commented on lib/elixir/test/elixir/string_test.exs:
```diff
@@ -195,10 +202,13 @@ defmodule StringTest do
     assert String.lstrip("猫猫 cat   ", ?猫) == " cat   "
+    # information separators are not whitespace 
+    assert String.lstrip("\u001F a  abc  a") == <<31>> <> " a  abc  a"
+    # no-break whitespace
```
And there `no-break whitespace` –> `no-break space`.
And one below. :bowtie:

lexmag commented on lib/elixir/unicode/unicode.ex:
```diff
@@ -12,11 +12,25 @@ defmodule String.Unicode do
+  prop_path = Path.join(__DIR__, "PropList.txt")
+  prop_range_line_regex = ~r/\A([0-9A-F]{4})(?:\.\.([0-9A-F]{4}))?/
+  whitespace = Enum.reduce File.stream!(prop_path), [], fn(line, acc) ->
+    case Regex.run(prop_range_line_regex, line) do
```
`capture: all_but_first` to remove the first matching subpattern.

amuino commented:
New round of commits pushed 🙂

lexmag commented on lib/elixir/unicode/unicode.ex:
```diff
@@ -12,11 +12,25 @@ defmodule String.Unicode do
+      [_, single] ->
+        [to_binary.(single) | acc]
+      [_, first, last] ->
+        [first_i, last_i] = [String.to_integer(first, 16), String.to_integer(last, 16)]
```
I think we can define right here:
```elixir
range = String.to_integer(first, 16)..String.to_integer(last, 16)
```

lexmag commented:
@amuino last comment about the `range` definition, otherwise it looks wonderful. :yellow_heart:

lexmag commented:
@amuino could you also squash everything into single commit?

amuino commented:
Rebased against master and squashed.

There is one additional test covering characters defined inside a range on `WhiteSpace.txt`, just in case.

@lexmag on top of the range, I also added your `capture: :all_but_first` suggestion

josevalim commented:
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:
